riscos_toolbox/gadgets/draggable.py

```diff
@@ -0,0 +1,94 @@
+"""RISC OS Toolbox - Gadgets - Draggable"""
+
+import ctypes
+import swi
+from . import Gadget, GadgetDefinition
+from ..events import ToolboxEvent
+
+class Draggable(Gadget):
+	_type = 640
+	
+	# Methods	           
+	SetSprite = _type + 0
+	GetSprite = _type + 1
+	SetText   = _type + 2
+	GetText   = _type + 3
+	SetState  = _type + 4
+	GetState  = _type + 5
+	
+	# Events
+	DragStarted = 0x82887 # Window_SWIChunkBase (0x82880) + 7
+	DragEnded   = 0x82888 # Window_SWIChunkBase (0x82880) + 8
+	
+	# Flags
+	GenerateDragStarted = 0x00000001
+	Sprite              = 0x00000002
+	TypeShift           = 0x00000003
+	Text                = 0x00000004
+	TypeMask            = 0x00000038
+	ToolboxIds          = 0x00000040
+	HasDropShadow       = 0x00000080
+	NotDithered         = 0x00000100
+	
+	@property
+	def sprite(self):
+		return self._miscop_get_text(Draggable.GetSprite)
+		
+	@sprite.setter
+	def sprite(self, sprite):
+		return self._miscop_set_text(Draggable.SetSprite, sprite)
+		
+	@property
+	def text(self):
+		return self._miscop_get_text(Draggable.GetText)
+		
+	@text.setter
+	def text(self, text):
+		return self._miscop_set_text(Draggable.SetText, text)
+		
+	@property
+	def state(self):
+		return self._miscop_get_int(self.GetState)
+		
+	@state.setter
+	def state(self, state):
+		return self._miscop_set_int(Draggable.SetState, state)
+	
+class DraggableDefinition(GadgetDefinition):
+	_gadget_class = Draggable
+	_fields_ = [ ("text", ctypes.c_char_p),
+	             ("max_text_len", ctypes.c_int32),
+	             ("sprite", ctypes.c_char_p),
+	             ("max_sprite_len", ctypes.c_int32) ]
+	            
+class DraggableDragStartedEvent(ToolboxEvent):
+	event_id = Draggable.DragStarted
+	
+	# This event actually doesn't have any fields other than the header
+	_fields_ = [ ]
+	
+class DraggableDragEndedEvent(ToolboxEvent):
+	event_id = Draggable.DragEnded
+	             	
+	 # Note that window handle and icon handle here can be object/component ID if
+	 # ToolboxIds flag is set
```

Comment:
Maybe add a property toolbox_id which returns a tuple of the object/component ID. If you want to be clever the state of ToolboxIds could be checked in this (and window_handle/icon_handle) and raise an exception if you use the "wrong" one.

Reply:
I'm not quite sure what you mean by the second point as far as using the wrong one, but on the first one, the window_handle and icon_handle can be object/component id if the ToolboxIds flag is set, so perhaps we could make this a single tuple property called more generically "ids" or something, and provide some kind of method in the Draggable class to easily flip that flag using the Gadget flags setter? Does that make sense?